wolfcrypt/src/wc_lms_impl.c

```diff
@@ -19,8 +19,3072 @@
+ * @return  0 when private key exhausted.
+ */
+int wc_hss_sigsleft(const LmsParams* params, const byte* priv_raw)
+{
+    w64wrapper q;
+    w64wrapper cnt;
+
+    /* Get current q - next leaf index to sign with. */
+    ato64(priv_raw, &q);
+    /* 1 << total_height = total leaf nodes. */
+    cnt = w64ShiftLeft(w64From32(0, 1), params->levels * params->height);
+    /* Check q is less than total leaf node count. */
+    return w64LT(q, cnt);
+}
+#endif /* !WOLFSSL_LMS_VERIFY_ONLY */
+
+/* Verify message using HSS.
+ *
+ * Section 6.3. Signature Verification
+ *  1. Nspk = strTou32(first four bytes of S)
+ *  2. if Nspk+1 is not equal to the number of levels L in pub:
+ *  3.   return INVALID
+ *  4. key = pub
+ *  5. for (i = 0; i < Nspk; i = i + 1) {
+ *  6.   sig = siglist[i]
+ *  7.   msg = publist[i]
+ *  8.   if (lms_verify(msg, key, sig) != VALID):
+ *  9.     return INVALID
+ * 10.   key = msg
+ * 11. }
+ * 12. return lms_verify(message, key, siglist[Nspk])
+ *
+ * @param [in, out] state  LMS state.
+ * @param [in]      pub    HSS public key.
+ * @param [in]      msg    Message to rifyn.
+ * @param [in]      msgSz  Length of message in bytes.
+ * @param [in]      sig    Signature of message.
+ * @return  0 on success.
+ * @return  SIG_VERFIY_E on failure.
+ */
+int wc_hss_verify(LmsState* state, const byte* pub, const byte* msg,
+    word32 msgSz, const byte* sig)
+{
+    const LmsParams* params = state->params;
+    int ret = 0;
+    word32 nspk;
+    const byte* key = pub + LMS_L_LEN;
+    word32 levels;
+
+    /* Get number of levels from public key. */
+    ato32(pub, &levels);
+    /* Line 1: Get number of signed public keys from signature. */
+    ato32(sig, &nspk);
+    /* Line 6 (First iteration): Move to start of next signature. */
+    sig += LMS_L_LEN;
+
+    /* Line 2: Verify that pub and signature match in levels. */
+    if (nspk + 1 != levels) {
+        /* Line 3: Return invalid signature. */
+        ret = SIG_VERIFY_E;
+    }
+    if (ret == 0) {
+        word32 i;
+
+        /* Line 5: For all but last LMS signature. */
+        for (i = 0; (ret == 0) && (i < nspk); i++) {
+            /* Line 7: Get start of public key in signature. */
+            const byte* pubList = sig + LMS_Q_LEN + LMS_TYPE_LEN +
+                LMS_MAX_NODE_LEN + params->p * LMS_MAX_NODE_LEN + LMS_TYPE_LEN +
+                params->height * LMS_MAX_NODE_LEN;
+            /* Line 8: Verify the LMS signature with public key as message. */
+            ret = wc_lms_verify(state, key, pubList, LMS_PUBKEY_LEN, sig);
+            /* Line 10: Next key is from signature. */
+            key = pubList;
+            /* Line 6: Move to start of next signature. */
+            sig = pubList + LMS_PUBKEY_LEN;
+        }
+    }
+    if (ret == 0) {
+        /* Line 12: Verify bottom tree with real message. */
+        ret = wc_lms_verify(state, key, msg, msgSz, sig);
+    }
+
+    return ret;
+}
+
+#endif /* WOLFSSL_HAVE_LMS && WOLFSSL_WC_LMS */
+
```

Comment:
I think blank line at end of file here and in wc_xmss_impl.c is causing test to fail.

Reply:
fixed